src/main/flight/pid.c

```diff
@@ -292,13 +295,97 @@ static void pidApplyRateController(const pidProfile_t *pidProfile, pidState_t *p
 #endif
 }
 
+void updateMagHoldHeading(int16_t heading)
+{
+    magHoldTargetHeading = heading;
+}
+
+int16_t getMagHoldHeading() {
+    return magHoldTargetHeading;
+}
+
+uint8_t getMagHoldState()
+{
+
+    #ifndef MAG
+        return MAG_HOLD_DISABLED;
+    #endif
+
+    if (!sensors(SENSOR_MAG) || !STATE(SMALL_ANGLE)) {
+        return MAG_HOLD_DISABLED;
+    }
+
+#if defined(NAV)
+    int navHeadingState = naivationGetHeadingControlState();
+    // NAV will prevent MAG_MODE from activating, but require heading control
+    if (navHeadingState != NAV_HEADING_CONTROL_NONE) {
+        // Apply maghold only if heading control is in auto mode
+        if (navHeadingState == NAV_HEADING_CONTROL_AUTO) {
+            return MAG_HOLD_ENABLED;
+        }
+    }
+    else
+#endif
+    if (ABS(rcCommand[YAW]) < 15 && FLIGHT_MODE(MAG_MODE)) {
+        return MAG_HOLD_ENABLED;
+    } else {
+        return MAG_HOLD_UPDATE_HEADING;
+    }
+}
+
+int16_t pidMagHold(const pidProfile_t *pidProfile)
+{
+
+    static filterStatePt1_t magHoldErrorFilter;
+
+    int16_t error = DECIDEGREES_TO_DEGREES(attitude.values.yaw) - magHoldTargetHeading;
+
+    if (error <= -180) {
+        error += 360;
+    }
+
+    if (error >= +180) {
+        error -= 360;
+    }
+
+    /*
+     * Try limiting diff. If big diff appeared that means that probably this is due to WAYPOINT or RTH fligh mode
+     * Too much diff might cause rapid yaw response and general UAV instability
+     */
+    error = constrain(error, (int) (-pidProfile->mag_hold_heading_diff_limit), pidProfile->mag_hold_heading_diff_limit);
+
+    //Apply LPF filter to smoothen response
+    error = filterApplyPt1(error, &magHoldErrorFilter, MAG_HOLD_ERROR_LPF_FREQ, dT);
+
+    //Let's scale it down and replace with something usefull
+    return error * pidProfile->P8[PIDMAG] / 30;
```

Comment:
To be honest, I was thinking slightly different, more flexible approach. 
Why use manual yaw rate for MAG_HOLD own rate? Why MAG_HOLD at maximum allowed output assumes more less half the rate of manual yaw?
It will be better to detach MAG_HOLD from manual input completely and let it live completely free from manual rates. I've been thinking about it for a while. Originally I wanted to do it in next step, but on the other hand, why not right now...